### PR TITLE
Fix DiesThisOrAnotherCreatureTriggeredAbility to work correctly with Devour

### PR DESCRIPTION
**Pull request by Dilnu**
Fixes https://github.com/magefree/mage/issues/6254

### CODE REVIEW

File: Mage/src/main/java/mage/abilities/common/DiesThisOrAnotherCreatureTriggeredAbility.java [hunk 1]
```diff
@@ -44,18 +44,8 @@ public boolean checkEventType(GameEvent event, Game game) {
...
-        } else {
-            if (game.getShortLivingLKI(getSourceId(), Zone.BATTLEFIELD)) {
-                sourcePermanent = (Permanent) game.getLastKnownInformation(getSourceId(), Zone.BATTLEFIELD);
-            }
-        }
-        if (sourcePermanent == null) {
-            return false;
-        }
-        return hasSourceObjectAbility(game, sourcePermanent, event);
+        Permanent sourcePermanent = game.getPermanentOrLKIBattlefield(getSourceId());
```
Comment (JayDi85) [inline Mage/src/main/java/mage/abilities/common/DiesThisOrAnotherCreatureTriggeredAbility.java]: That radical changes with removes of `getShortLivingLKI` need more research and tests. That's code uses in many another abilities too. Maybe @LevelX2 can hint on that.

**Comment by Dilnu** [comment]
I've tried several ways to fix this. My current solution isn't perfect either because it breaks certain stacked replacement effects as shown by the DesertionTest.

I'd love any advice that's available.

The core of the bug is that shortLKI is cleared while moving the Creature with devour onto the battlefield which means it isn't available while handling the events that move the sacrifices creatures.

**Comment by Dilnu** [comment]
Apparently I spoke too soon. I tried the simplest fix of simply not clearing shortLKI at the place where clearing shortLKI was causing issues and that worked without breaking any other tests.

**Comment by JayDi85** [comment]
The main problem: xmage's tests cover percent is... 0.5-2% That's why if you change something and tests works -- it's not guaranty that it works fine. What's why developers must research what these changes could potentially affect.

**Comment by Dilnu** [comment]
shortLivingLKI is used to determine whether two zone changes occurred at the same time.

getShortLivingLKI is only used in "isInUsableZone" and in two cases which may be technically incorrect in "checkTrigger" in either case it is used to determine whether triggered abilities trigger or static abilities come into effect.

Prior to this change if cards moved to multiple zones and one of those zones was the battlefield then we clear the LKI prematurely as seen in the case of devour. I think it's also worth noting that we were not doing this for moving cards to any other zone.

It seems fairly clear to me that the reason for the line. `game.applyEffects()` was for the portion of the function that I've retained. That portion of the ability is required for effects like Blood Moon to function correctly because those effects need to be applied before we check to see if abilities trigger on entering the battlefield and similar situations.

I also think that this is in fact one of the parts of the code base which has high test coverage. There's a lot of tests that failed for every other change that I attempted to moveCards.

I am confident that the new behavior is correct.

**Comment by jeffwadsworth** [comment]
I guess this is one of those things that remains in limbo for years.

**Comment by Dilnu** [comment]
I did the research I was asked to do... maybe it would make sense to ask LevelX2's opinion?

**Comment by JayDi85** [comment]
Real fix was made by 85083a4433a25b07fd4b9340b1c195cc8e32f426, I also copy test and code improve from that PR too.

There is must be one line fix for multi step effects (see original commit d8fb962030febe3834810702d120943ce1ff69ca by @LevelX2 with `game.applyEffects()` replacement with new one):
* `game.getState().processAction(game);`

**Pull request closed without merging**